src/hierarchies/lapnig_hierarchy.cc

```diff
@@ -0,0 +1,209 @@
+  if (add) {
+
+    cluster_data_values.push_back(datum); // add the datum to cluster_data_values, needed to compute the full_cond
+
+  } else {
+
+    auto it = std::find(cluster_data_values.begin(), cluster_data_values.end(), datum);
+    cluster_data_values.erase(it);
+
+  }
+}
+
+
+void LapNIGHierarchy::sample_full_cond(bool update_params) {
+
+  if (this->card == 0) {
+      // No posterior update possible
+      this->sample_prior();
+  } else {
+      // Number of iterations to compute the acceptance rate
+      ++iter_;
+
+      // Random generator
+      auto &rng = bayesmix::Rng::Instance().get();
+
+      // Candidate mean and candidate log_scale
+      double candidate_mean_step_var{2*hypers->var}; // 0.5 * Variance of the mean distribution
+
+      double candidate_log_scale_step_var{2*boost::math::trigamma(2*hypers->shape)};  // 0.5 * Variance of log-gamma distribution
+
+      double candidate_mean = state.mean + stan::math::normal_rng(0,sqrt(candidate_mean_step_var), rng);
+
+      double candidate_log_scale = std::log(state.scale) +
+                                   stan::math::normal_rng(0,sqrt(candidate_log_scale_step_var), rng);
+
+      double candidate_scale = std::exp(candidate_log_scale);
+
+      //MH step
+
+      double pi_current, pi_candidate; // posterior of current state and candidate state
+
+      double candidate_sum = 0; // Sum of absolute values of data - candidate_mean
+      for(auto & elem: cluster_data_values){
+        candidate_sum += std::abs(elem(0,0) - candidate_mean); // MODIFY IN THE FUTURE FOR MULTIVARIATE CASE!!!
+      }
+
+      pi_candidate = 0.5/candidate_scale *
+                std::exp(-0.5 / candidate_scale * candidate_sum) *
+                std::exp(-0.5 / hypers->var / hypers->var *
+                         (candidate_mean - hypers->mean) *
+                         (candidate_mean - hypers->mean)) *
+                std::pow(1./ candidate_scale, hypers->shape) *
+                std::exp(-1./hypers->scale / candidate_scale);
+
+      double current_sum = 0; // Sum of absolute values of data - candidate_mean
+
+      double current_mean = state.mean;
+      double current_scale = state.scale;
+
+      for(auto & elem: cluster_data_values){
+        current_sum += std::abs(elem(0,0) - current_mean); // MODIFY IN THE FUTURE FOR MULTIVARIATE CASE!!!
+      }
+      pi_current = 0.5/current_scale * std::exp(-0.5/current_scale*current_sum) *
+                std::exp(-0.5/hypers->var/hypers->var
+                         *(current_mean-hypers->mean)*(current_mean-hypers->mean))*
+                std::pow(1./current_scale,hypers->shape) * std::exp(-1./hypers->scale/current_scale);
+
+      double alpha;
+      if(pi_current!=0) {
+        alpha = std::min<double>((pi_candidate/pi_current), 1);
+      }
+      else{
+        alpha = 1;
+      }
+
+      //std::cout << "alpha: " << alpha << std::endl;
+
+      bool accept = stan::math::bernoulli_rng(alpha,rng);
```

Comment:
This code is rather complex to read and I think could be improved.
First, you could divide it into several functions:

```
Eigen::VectorXd propose_rwmh(Eigen::VectorXd curr_vals) {} --> samples from the proposal distribution

double eval_prior_lpdf_unconstrained(Eigen::VectorXd unconstrained_parameters); --> evaluates the prior given the mean and log of the scale (beware of the change of variables!

double eval_like_lpdf_unconstrained(Eigen::VectorXd unconstrained_parameters); --> evaluates the (sum of the) log likelihood for all the observations in the cluster
```

Then in the current function you could have something like
```
Eigen::VectorXd curr_unc_params(2)
curr_unc_params <<  state.mean, std::log(state.scale);

Eigen::VectorXd prop_unc_params = propose_rwmh(curr_unc_params);

double log_target_prop = eval_prior_lpdf_unconstrained(prop_unc_params) + eval_like_lpdf_unconstrained(prop_unc_params);

double log_target_curr = eval_prior_lpdf_unconstrained(curr_unc_params) + eval_like_lpdf_unconstrained(curr_unc_params);

double log_arate = log_target_prop - log_target_curr;

if (std::log(stan::math::uniform_rng(0, 1, rng) < log_arate) {
  state.mean = prop_unconstrained(0);
  state.scale = std::exp(prop_unconstrained(1);
}
---
```